sequencing/transaction_queue.go

```diff
@@ -0,0 +1,181 @@
+package sequencing
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/dgraph-io/badger/v4"
+
+	"github.com/rollkit/go-sequencing"
+)
+
+// BatchExtender is an interface for extending a batch of transactions
+type BatchExtender interface {
+	Head(max uint64) ([]byte, error)
+	Tail(max uint64) ([]byte, error)
+}
+
+// TransactionQueue is a queue of transactions
+type TransactionQueue struct {
+	queue    []sequencing.Tx
+	mu       sync.Mutex
+	extender BatchExtender
+}
+
+// NewTransactionQueue creates a new TransactionQueue
+func NewTransactionQueue(extender BatchExtender) *TransactionQueue {
+	return &TransactionQueue{
+		queue:    make([]sequencing.Tx, 0),
+		extender: extender,
+	}
+}
+
+// GetTransactionHash to get hash from transaction bytes using SHA-256
+func GetTransactionHash(txBytes []byte) string {
+	hashBytes := sha256.Sum256(txBytes)
+	return hex.EncodeToString(hashBytes[:])
+}
+
+// AddTransaction adds a new transaction to the queue
+func (tq *TransactionQueue) AddTransaction(tx sequencing.Tx, db *badger.DB) error {
+	tq.mu.Lock()
+	tq.queue = append(tq.queue, tx)
+	tq.mu.Unlock()
+
+	// Store transaction in BadgerDB
+	err := db.Update(func(txn *badger.Txn) error {
+		key := append(keyPrefixTx, []byte(GetTransactionHash(tx))...)
+		return txn.Set(key, tx)
+	})
+	return err
+}
+
+// GetNextBatch extracts a batch of transactions from the queue
+func (tq *TransactionQueue) GetNextBatch(max uint64, db *badger.DB) sequencing.Batch {
+	tq.mu.Lock()
+	defer tq.mu.Unlock()
+
+	// Add head of batch if extender is provided, also ask for the tail of the batch
+	var head, tail []byte
+	if tq.extender != nil {
+		var err error
+		head, err = tq.extender.Head(max)
+		if err != nil {
+			return sequencing.Batch{Transactions: nil}
+		}
+		tail, err = tq.extender.Tail(max)
+		if err != nil {
+			return sequencing.Batch{Transactions: nil}
+		}
+	}
+
+	batch := tq.queue
+	headTailSize := len(head) + len(tail)
+
+	for uint64(totalBytes(batch)+headTailSize) > max {
+		batch = batch[:len(batch)-1]
+	}
```

Comment:
This is going to arbitrarily cut bytes off of txns in the queue which will corrupt the txns. 
Instead of truncation, the for loop should be removing entire txns from the queue until the batch is under the max.

A future optimization of the batch would be to make sure the txns in the queue are ordered by a priority. The priority could be customized to either be size, gas, timestamp etc. 

For this implementation though I wouldn't worry about the queue ordering as how to prioritize txns should be a separate discussion.